hk2-locator/src/main/java/org/jvnet/hk2/internal/ServiceLocatorImpl.java

```diff
@@ -2002,19 +2002,17 @@ private void reupInstanceListenersHandlers(Collection<SystemDescriptor<?>> check
     private void reupClassAnalyzers() {
         List<ServiceHandle<?>> allAnalyzers = protectedGetAllServiceHandles(ClassAnalyzer.class);
 
-        synchronized (classAnalyzerLock) {
```

Comment:
Hmmm, right, if I don't have absolute control over the service, it is pointless to wrap it in sync block.

Reply:
`reupClassAnalyzers` method invoked only in one place under `ReentrantReadWriteLock.WriteLock`

Reply:
The main idea - get rid of `classAnalyzerLock` and use `wLock` and `rLock` instead. This prevent us from using lifecycle methods while (re)configuration is in progress.